lms/djangoapps/courseware/masquerade.py

```diff
@@ -99,6 +126,29 @@ def get_masquerade_role(user, course_key):
     return course_masquerade.role if course_masquerade else None
 
 
+def get_masquerade_user(user, course_key):
+    """
+    Returns the specific Django user the authenticated user is masquerading as, or None if masquerading as a
+    specific user is not active or invalid.
+    """
+    # TODO(smarnach): We should notify the caller if the user_name in the masquerading settings don't refer
+    # to a valid user enrolled in this course.  This shouldn't normally be possible, since we check whether the
+    # user is enrolled in the AJAX callback, but the enrollment state of a user might have changed in the
+    # meantime.  We could also reset the masquerading settings in that case.
+    course_masquerade = get_course_masquerade(user, course_key)
+    if course_masquerade and course_masquerade.user_name:
+        try:
+            masquerade_user = CourseEnrollment.objects.users_enrolled_in(course_key).get(username=course_masquerade.user_name)
+        except User.DoesNotExist:
+            pass
+        else:
+            masquerade_user.masquerade_settings = user.masquerade_settings
+            del user.masquerade_settings
```

Comment:
Masquerading isn't implemented transparently – the masquerading settings are checked explicitly in different places of the code.  One example is the [check in `has_access()` that explicitly calls `is_masquerading_as_student()`](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/access.py#L614).  If I would leave the masquerade settings attached to the `real_user` object, that user still wouldn't be granted staff access.  The `real_user` is used to decide whether to show the "Staff Debug Info" link, and leaving the masquerade settings in place would result in the link not showing up.  This solution seemed the easiest way to achieve the desired behaviour.

I added a comment with a bit of explanation.